File: api-ruby/samples/server/direct.rb
require 'sinatra/base'
require_relative 'config'
module DirectRoutes
  def self.registered(app)
    app.get '/direct' do
      send_file File.join(settings.public_folder, 'direct.html')
    end

    app.get '/direct-popup' do
      send_file File.join(settings.public_folder, 'direct-popup.html')
    end

    app.post('/create-direct-session/:provider') do
      provider = params[:provider]
      redirectUrl = params[:redirectUrl]
      fallbackToTrinsicUI = params[:fallbackToTrinsicUI] == "true"
      capabilities = params[:capabilities].split(",")
      verificationProfileId = ENV['TRINSIC_VERIFICATION_PROFILE_ID']

      req = TrinsicApi::CreateDirectProviderSessionRequest.new({
        redirect_url: redirectUrl,
        provider: provider,
        capabilities: capabilities,
        fallback_to_hosted_ui: fallbackToTrinsicUI,
        verification_profile_id: verificationProfileId
      })

      opts = {
        create_direct_provider_session_request: req
      }

      result = TrinsicServices::SESSIONS.create_direct_provider_session(opts)
      json result.to_body
    end

    app.post('/refresh-content/:sessionId') do
      sessionId = params[:sessionId]
      request_payload = JSON.parse(request.body.read)
      results_access_key = request_payload["resultsAccessKey"]

      result = TrinsicServices::SESSIONS.refresh_step_content(sessionId, {
        refresh_step_content_request: TrinsicApi::RefreshStepContentRequest.new({
          results_access_key: results_access_key
        })
      })
      json DirectRoutes.deep_transform(result.to_body)
    end

    app.post('/poll-results/:sessionId') do
      request_payload = JSON.parse(request.body.read)
      # Accessing a specific parameter
      results_access_key = request_payload["resultsAccessKey"]

      result = TrinsicServices::SESSIONS.get_session_result(params[:sessionId], {
        get_session_result_request: TrinsicApi::GetSessionResultRequest.new({
          results_access_key: results_access_key
        })
      })

      json DirectRoutes.deep_transform(result.to_body)
    end

  end

  # We want to deeply transform as all Time values should be changed to iso8601 to be compatible with our web ui SDK
  def self.deep_transform(obj)
    case obj
    when Hash
      obj.transform_values { |v| deep_transform(v) }
    when Array
      obj.map { |v| deep_transform(v) }
    when Time
      obj.iso8601
    else
      obj
    end
  end

end

File: api-python/samples/server/direct.py
from fastapi import Request, APIRouter, Depends, HTTPException
from fastapi.responses import JSONResponse
from trinsic_api.api.sessions_api import SessionsApi, CreateDirectProviderSessionRequest, GetSessionResultRequest, RefreshStepContentRequest
import os
from urllib.parse import urlencode
from datetime import datetime, timezone
from fastapi.responses import RedirectResponse
import json
from datetime import date, datetime

directRouter = APIRouter()

@directRouter.get("/direct")
async def direct():
    return RedirectResponse(url="/direct.html")

@directRouter.get("/direct-popup")
async def directPopup(request: Request):
    return RedirectResponse(url=f"/direct-popup.html?{request.query_params}")


@directRouter.post("/create-direct-session/{provider_id}")
async def launch(request: Request, provider_id: str, sessions_api: SessionsApi = Depends()):
    fallbackToTrinsicUI = request.query_params.get("fallbackToTrinsicUI", "").strip().lower() == "true"

    request = CreateDirectProviderSessionRequest(redirect_url = request.query_params.get("redirectUrl"), verification_profile_id=os.getenv("TRINSIC_VERIFICATION_PROFILE_ID"), provider=provider_id, fallback_to_hosted_ui = fallbackToTrinsicUI, capabilities = request.query_params.get("capabilities").split(","))

    result = sessions_api.create_direct_provider_session(create_direct_provider_session_request=request)
    return JSONResponse(content=json.loads(json.dumps(result.to_dict(), default=json_serial)))

def json_serial(obj):
    """JSON serializer for objects not serializable by default."""
    if isinstance(obj, (datetime, date)):
        return obj.isoformat()
    raise TypeError(f"Type {type(obj)} not serializable")

@directRouter.post("/poll-results/{session_id}")
async def poll_results(session_id: str, request: Request, sessions_api: SessionsApi = Depends()):
    request_payload = await request.json()
    results_access_key = request_payload.get("resultsAccessKey")
    
    if not results_access_key:
        raise HTTPException(status_code=400, detail="Missing resultsAccessKey")
    
    result = sessions_api.get_session_result(session_id, GetSessionResultRequest(results_access_key=results_access_key))
    
    return JSONResponse(content=json.loads(json.dumps(result.to_dict(), default=json_serial)))

@directRouter.post("/refresh-content/{session_id}")
async def refresh_content(session_id: str, request: Request, sessions_api: SessionsApi = Depends()):
    request_payload = await request.json()
    results_access_key = request_payload.get("resultsAccessKey")
    
    if not results_access_key:
        raise HTTPException(status_code=400, detail="Missing resultsAccessKey")
    
    result = sessions_api.refresh_step_content(session_id, RefreshStepContentRequest(results_access_key=results_access_key))
    
    return JSONResponse(content=json.loads(json.dumps(result.to_dict(), default=json_serial)))